feat(ui-components): add states and full-width support to Button component

huyenltnguyen opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Ref #45357

<!-- Feel free to add any additional description of changes below this line -->

This PR adds the following support to the Button component:
- Active state
- Focus state
- Disabled state (via the `disabled` prop)
- Full-width usage (via the `block` prop)

Note: The PR can be reviewed by commits.

## Screenshots

### Active state

I'm adding a screenshot of the normal state for comparison:

| Normal | Active |
| --- | --- |
| <img width="104" alt="Screen Shot 2022-04-11 at 01 47 14" src="https://user-images.githubusercontent.com/25715018/162634876-5f923054-9622-48bf-a41c-123ecaac893f.png"> | <img width="104" alt="Screen Shot 2022-04-11 at 01 21 24" src="https://user-images.githubusercontent.com/25715018/162634884-6bef4714-ec36-43d2-ad60-30d5ce99be1b.png"> |

### Focus state

| | Light | Dark |
| --- | --- | --- |
| Primary | <img width="109" alt="Screen Shot 2022-04-11 at 14 46 28" src="https://user-images.githubusercontent.com/25715018/162689220-45358127-1b6d-4065-a13c-4bbe871a5520.png"> | <img width="105" alt="Screen Shot 2022-04-11 at 14 46 35" src="https://user-images.githubusercontent.com/25715018/162689242-932f6166-2bd0-4a40-a37e-5e74c1d9088f.png"> |
| Danger | <img width="107" alt="Screen Shot 2022-04-11 at 14 43 44" src="https://user-images.githubusercontent.com/25715018/162689294-d46499d3-c444-4c37-bf86-3042b8f1a5c9.png"> | <img width="109" alt="Screen Shot 2022-04-11 at 14 47 23" src="https://user-images.githubusercontent.com/25715018/162689345-3622dd48-c383-4a00-b2e5-40098a2e95b5.png"> |
| Info | <img width="107" alt="Screen Shot 2022-04-11 at 14 47 57" src="https://user-images.githubusercontent.com/25715018/162689457-77e3d3c0-2936-479e-b34d-a56185d43bbc.png"> | <img width="106" alt="Screen Shot 2022-04-11 at 14 47 51" src="https://user-images.githubusercontent.com/25715018/162689476-3bb9a8b8-ebd4-4b11-ab96-8b112d10dd35.png"> |

### Disabled state

Note: Inactive/disabled elements are allowed to not meet the contrast minimum (Ref: https://www.w3.org/TR/WCAG21/#contrast-minimum)

| | Light | Dark |
| --- | --- | --- |
| Primary | <img width="104" alt="Screen Shot 2022-04-11 at 01 21 13" src="https://user-images.githubusercontent.com/25715018/162634925-4cfba4dd-b67e-423f-a163-46b9212aee3b.png"> | <img width="104" alt="Screen Shot 2022-04-11 at 01 50 07" src="https://user-images.githubusercontent.com/25715018/162634997-a921a032-c7ac-4e80-b59a-2a10932cf744.png"> |
| Danger | <img width="97" alt="Screen Shot 2022-04-11 at 01 48 58" src="https://user-images.githubusercontent.com/25715018/162635125-bd2103ee-8c7d-4121-af7b-717c2b09922a.png"> | <img width="103" alt="Screen Shot 2022-04-11 at 01 53 24" src="https://user-images.githubusercontent.com/25715018/162635119-3606e24f-31de-4a4d-a4a6-b183c584ceec.png"> |
| Info | <img width="101" alt="Screen Shot 2022-04-11 at 01 54 07" src="https://user-images.githubusercontent.com/25715018/162635156-2d03316e-926e-4d0a-983a-bc726ca91a7d.png"> | <img width="101" alt="Screen Shot 2022-04-11 at 01 53 59" src="https://user-images.githubusercontent.com/25715018/162635152-06fa889a-cac6-4859-b9bd-996999795a3c.png"> |

### Full-width button

<img width="924" alt="Screen Shot 2022-04-11 at 01 28 08" src="https://user-images.githubusercontent.com/25715018/162635181-19851ba9-ed46-48e9-9cac-16910257c195.png">

huyenltnguyen committed:
add active style

huyenltnguyen committed:
add disable state

huyenltnguyen committed:
add support for full-width

huyenltnguyen commented on tools/ui-components/src/button/button.tsx:
```diff
@@ -1,35 +1,66 @@
         'text-default-foreground-danger',
-        'hover:bg-default-background-danger-hover',
-        'hover:text-default-foreground-danger-hover'
+        ...(disabled
+          ? []
+          : [
+              'hover:bg-default-background-danger-hover',
+              'hover:text-default-foreground-danger-hover'
+            ])
```
This would be `&:hover:not([aria-disabled="true"])` if we were to write in CSS. Unfortunately Tailwind doesn't allow a `not` selector so I had to add a condition here.

huyenltnguyen commented on tools/ui-components/src/button/button.tsx:
```diff
@@ -65,17 +100,39 @@ export const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
+        className={classes}
+        type={type}
+        onClick={handleClick}
+        aria-disabled={disabled}
```
`aria-disabled` is more preferred as it allows the user to focus on the element and understand that the button is disabled. While with the `disabled` attribute, the button is not focusable, and non-sighted users won't know about its existence nor its disabled state.

Refs: 
- https://www.a11y-101.com/development/aria-disabled
- https://css-tricks.com/making-disabled-buttons-more-inclusive

huyenltnguyen commented on tools/ui-components/tailwind.config.js:
```diff
@@ -37,6 +39,17 @@ module.exports = {
       'default-background-danger-hover':
         'var(--default-background-danger-hover)',
       'default-background-info-hover': 'var(--default-background-info-hover)',
+      gray: {
+        0: 'var(--gray00)',
+        50: 'var(--gray05)',
+        100: 'var(--gray10)',
+        150: 'var(--gray15)',
+        450: 'var(--gray45)',
+        750: 'var(--gray75)',
+        800: 'var(--gray80)',
+        850: 'var(--gray85)',
+        900: 'var(--gray90)'
+      },
```
I only need `gray90`, but adding all of the variables in anyway, since we already did that for other colors.

bbsmooth commented:
> * I did not cover the focus state as browsers automatically add a focus ring to the element on focus, which is sufficient for a11y support.

You are correct, it is technically sufficient for now under WCAG 2.1 but it won't be under WCAG 2.2 as [SC 2.4.11](https://www.w3.org/WAI/WCAG22/Understanding/focus-appearance-minimum.html) is being added which will create a new minimum standard for focus appearance, which many browsers will not meet, so basically we will need to make sure we have a good focus custom focus indicator that can be seen in all browsers. In PR #45137 I suggested we use `--blue-mid` as the focus ring color as it contrasts nicely in both the light and dark themes.

WCAG 2.2 probably won't be officially released until this Fall but I don't think that should stop us from having a better focus indicator now.

huyenltnguyen commented on tools/ui-components/tailwind.config.js:
```diff
@@ -91,5 +104,9 @@ module.exports = {
     }
   },
-  plugins: []
+  plugins: [
+    plugin(({ addVariant }) => {
+      addVariant('aria-disabled', '[aria-disabled="true"]');
+    })
+  ]
```
Tailwind supports a lot of selectors by default ([ref](https://tailwindcss.com/docs/hover-focus-and-other-states#quick-reference)), but `aria-disabled` isn't one of them. I followed [this guide](https://tailwindcss.com/docs/hover-focus-and-other-states#creating-custom-modifiers) to add our own modifier. This allows us to style the disabled state using `aria-disabled:...` classes.

And at this point, I have to say I'm very impressed with Tailwind. They support everything!!! And their documentation is very well-written.

huyenltnguyen committed:
add custom focus outline

huyenltnguyen commented:
@bbsmooth Thanks a lot for the guidance!

I vaguely remember reading about this new requirement in the past, but I wasn't aware of the date it would be applied. (And it being a draft made me think it would be a while until the internet elders decide to roll it out lol.)

I have updated the PR to use our own focus ring and I have also added screenshots to the PR description. Let me know what you think 🙂

ahmadabdolsaheb commented:
Thanks again for the detailed notes. My only nitpick is that the disabled button looks like it is registering a click when it is not. 
![clickable](https://user-images.githubusercontent.com/4591597/163780202-ea7f9c52-f33e-415a-aeb8-197385a2137b.gif)

huyenltnguyen commented:
Ouch! Thanks Ahmad, that's a nice catch! 

This is probably a good opportunity for new contributors to get familiar with our UI components. I'll create a issue for this 🙂
